query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs

```diff
@@ -63,113 +62,79 @@ impl ChainingHashTable {
-            .map(|expr| {
-                ExpressionEvaluator::eval(self.ctx.try_get_function_context()?, expr, &input)
-            })
+            .map(|expr| ExpressionEvaluator::eval(&func_ctx, expr, &input))
             .collect::<Result<Vec<ColumnRef>>>()?;
 
-        let hash_values = self.hash(&build_keys, input.num_rows())?;
-
-        self.row_space.push(input, hash_values)?;
+        let serialized_build_keys = self.serialize_keys(&build_keys, input.num_rows())?;
+        self.row_space.push(input, serialized_build_keys)?;
 
         Ok(())
     }
 
     fn probe(&self, input: &DataBlock) -> Result<Vec<DataBlock>> {
+        let func_ctx = self.ctx.try_get_function_context()?;
         let probe_keys = self
             .probe_expressions
             .iter()
-            .map(|expr| {
-                ExpressionEvaluator::eval(self.ctx.try_get_function_context()?, expr, input)
-            })
+            .map(|expr| ExpressionEvaluator::eval(&func_ctx, expr, input))
             .collect::<Result<Vec<ColumnRef>>>()?;
-
+        let serialized_probe_keys = self.serialize_keys(&probe_keys, input.num_rows())?;
         let hash_table = self.hash_table.read().unwrap();
-        let hash_values = self.hash(&probe_keys, input.num_rows())?;
-        let hash_values =
-            ChainingHashTable::apply_bitmask(&hash_values, (hash_table.len() - 1) as u64);
 
         let mut results: Vec<DataBlock> = vec![];
-        for (i, hash_value) in hash_values.iter().enumerate().take(input.num_rows()) {
-            let probe_result_ptrs = hash_table[*hash_value as usize].as_slice();
-            if probe_result_ptrs.is_empty() {
+        for (i, key) in serialized_probe_keys
+            .iter()
+            .enumerate()
+            .take(input.num_rows())
+        {
+            let keys_ref = KeysRef::create(key.as_ptr() as usize, key.len());
+            let probe_result_ptr = hash_table.find_key(&keys_ref);
+            if probe_result_ptr.is_none() {
                 // No matched row for current probe row
                 continue;
             }
-            let result_block = self.row_space.gather(probe_result_ptrs)?;
-
+            let probe_result_ptrs = probe_result_ptr.unwrap().get_value();
+            // `result_block` is the block of build table
+            let result_blocks = self.row_space.gather(probe_result_ptrs)?;
             let probe_block = DataBlock::block_take_by_indices(input, &[i as u32])?;
-            let mut replicated_probe_block = DataBlock::empty();
-            for (i, col) in probe_block.columns().iter().enumerate() {
-                let replicated_col = ConstColumn::new(col.clone(), result_block.num_rows()).arc();
 
-                replicated_probe_block = replicated_probe_block
-                    .add_column(replicated_col, probe_block.schema().field(i).clone())?;
+            for result_block in result_blocks.iter() {
+                assert_eq!(result_block.clone().num_rows(), 1);
```

Comment:
Why does result_block only have 1 row of data?

Reply:
I changed the logic of `gather` so that instead of having `concat_blocks` in gather, it just returns `Vec<DataBlock>`. So now `result_block` has only one line, the same number of rows as `probe_block`.

Reply:
It looks wierd, take this example:

```
build side:
column1 column2
1       2
1       3
1       4

probe side:
column1
1
```
There is only one block of build side, i.e. chunks number is 1.

When you are probing with the probe key `1`, it should return all the 3 rows of data. In this case, how to make sure that `result_block` has only one row?

Reply:
I think the following code can split the single chunk to three chunks
```rust
            for row_ptr in row_ptrs.iter() {
                assert!((row_ptr.chunk_index as usize) < chunks.len());
                let block = self.gather_single_chunk(&chunks[row_ptr.chunk_index as usize], &[
                    row_ptr.row_index,
                ])?;
                if !block.is_empty() {
                    data_blocks.push(block);
                }
            }
```
The debug result of your case
```
[query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs:99] input =
[
    "+-------+-------+",
    "| \"c\"_0 | \"d\"_1 |",
    "+-------+-------+",
    "| 1     | 2     |",
    "+-------+-------+",
]

[query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs:125] result_blocks.clone() = [

    [
        "+-------+-------+",
        "| \"b\"_3 | \"a\"_2 |",
        "+-------+-------+",
        "| 2     | 1     |",
        "+-------+-------+",
    ]
    ,

    [
        "+-------+-------+",
        "| \"b\"_3 | \"a\"_2 |",
        "+-------+-------+",
        "| 3     | 1     |",
        "+-------+-------+",
    ]
    ,

    [
        "+-------+-------+",
        "| \"b\"_3 | \"a\"_2 |",
        "+-------+-------+",
        "| 4     | 1     |",
        "+-------+-------+",
    ]
    ,
]
[query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs:127] result_block.clone() =
[
    "+-------+-------+",
    "| \"b\"_3 | \"a\"_2 |",
    "+-------+-------+",
    "| 2     | 1     |",
    "+-------+-------+",
]

[query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs:127] result_block.clone() =
[
    "+-------+-------+",
    "| \"b\"_3 | \"a\"_2 |",
    "+-------+-------+",
    "| 3     | 1     |",
    "+-------+-------+",
]

[query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs:127] result_block.clone() =
[
    "+-------+-------+",
    "| \"b\"_3 | \"a\"_2 |",
    "+-------+-------+",
    "| 4     | 1     |",
    "+-------+-------+",
]
```